### ISSUE TITLE
I am trying to figure out how the model structure modification works. How can I exactly modify the structure and make it print out the new layers while training?

### ISSUE THREAD

**Issue by jinyoonok2** [ultralytics/ultralytics#7205]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, ultralytics team, I appreciate all the works for YOLOv8 and I understand it would be busy to maintain and develop YOLOv8 features.

But I am currently trying to figure out how can I modify YOLOv8 model structures by testing it out with brief changes.

I checked the related [issue](https://github.com/ultralytics/ultralytics/issues/1682):
1. If we make new classes in conv.py and block.py files.
2. then you have to change init.py and tasks.py files to import classes made by yourself.
3. Finally, you can add your new layer to the yolov8.yaml file.

and tried to follow up the instructions, but it does not seem to work.

I am currently using python 3.11 and running ultralytics on the virtual environment of PyCharm

**1. To start with ultralytics/nn/modules/conv.py and ultralytics/nn/modules/block.py**

![fakeconv](https://github.com/ultralytics/ultralytics/assets/84653106/562896d7-fe56-4f34-a8fc-f01522359a1e)

This is added in conv.py and

![fakeblock](https://github.com/ultralytics/ultralytics/assets/84653106/957664b5-6e71-48e1-88a5-c58583cf853d)

This is added in block.py

I just created fake conv and fake block that just passes the values to the next layer.
Purpose of these blocks and conv layer is that I just wanted to check if I can actually modify the layer structures.

**2. since now I created these, I change ultralytics/nn/modules/__init__.py and ultralytics/nn/tasks.py**

![init](https://github.com/ultralytics/ultralytics/assets/84653106/9f73aef8-4172-4c75-9faa-7e6a7aebf3e5)
Now I added them to __init__ file first,

![task_import](https://github.com/ultralytics/ultralytics/assets/84653106/96cdbddf-e28a-4b5d-8198-4d0f260a59c7)
![tasks](https://github.com/ultralytics/ultralytics/assets/84653106/6792c480-c54e-44cf-aacd-ad79ca37e549)

Then I imported it into **tasks.py** and make it inside the **'def parse_model(d, ch, verbose=True)' function** which utilizes the arguments from yaml model file and convert it into nn.Sequential to return them.

**below is the full code of the function parse_model, I only changed the parts where I put on the above picture**
`def parse_model(d, ch, verbose=True):  # model_dict, input_channels(3)
    """Parse a YOLO model.yaml dictionary into a PyTorch model."""
    import ast

    # Args
    max_channels = float('inf')
    nc, act, scales = (d.get(x) for x in ('nc', 'activation', 'scales'))
    depth, width, kpt_shape = (d.get(x, 1.0) for x in ('depth_multiple', 'width_multiple', 'kpt_shape'))
    if scales:
        scale = d.get('scale')
        if not scale:
            scale = tuple(scales.keys())[0]
            LOGGER.warning(f"WARNING ⚠️ no model scale passed. Assuming scale='{scale}'.")
        depth, width, max_channels = scales[scale]

    if act:
        Conv.default_act = eval(act)  # redefine default activation, i.e. Conv.default_act = nn.SiLU()
        if verbose:
            LOGGER.info(f"{colorstr('activation:')} {act}")  # print

    if verbose:
        LOGGER.info(f"\n{'':>3}{'from':>20}{'n':>3}{'params':>10}  {'module':<45}{'arguments':<30}")
    ch = [ch]
    layers, save, c2 = [], [], ch[-1]  # layers, savelist, ch out
    for i, (f, n, m, args) in enumerate(d['backbone'] + d['head']):  # from, number, module, args
        m = getattr(torch.nn, m[3:]) if 'nn.' in m else globals()[m]  # get module
        for j, a in enumerate(args):
            if isinstance(a, str):
                with contextlib.suppress(ValueError):
                    args[j] = locals()[a] if a in locals() else ast.literal_eval(a)

        # note:
        # f: from what layer, n: number of this layer, m: using module layer, args: arguments (last bracket input - layer size)
        # ch: list of layers from 1 to 22
        # c1: channel in ch[-1 or some other layer], c2: channel out [64, 3, 2] 64 here
        # ex args: [-1, 1, Conv, [64, 3, 2]]
        n = n_ = max(round(n * depth), 1) if n > 1 else n  # depth gain
        if m in (Classify, Conv, ConvTranspose, GhostConv, Bottleneck, GhostBottleneck, SPP, SPPF, DWConv, Focus,
                 BottleneckCSP, C1, C2, C2f, C3, C3TR, C3Ghost, nn.ConvTranspose2d, DWConvTranspose2d, C3x, RepC3, FakeConv, ModifiedFakeBlock):
            c1, c2 = ch[f], args[0]
            if c2 != nc:  # if c2 not equal to number of classes (i.e. for Classify() output)
                c2 = make_divisible(min(c2, max_channels) * width, 8)

            args = [c1, c2, *args[1:]]
            if m in (BottleneckCSP, C1, C2, C2f, C3, C3TR, C3Ghost, C3x, RepC3):
                args.insert(2, n)  # number of repeats
                n = 1
        elif m is AIFI:
            args = [ch[f], *args]
        elif m in (HGStem, HGBlock):
            c1, cm, c2 = ch[f], args[0], args[1]
            args = [c1, cm, c2, *args[2:]]
            if m is HGBlock:
                args.insert(4, n)  # number of repeats
                n = 1
        elif m is ResNetLayer:
            c2 = args[1] if args[3] else args[1] * 4
        elif m is nn.BatchNorm2d:
            args = [ch[f]]
        elif m is Concat:
            c2 = sum(ch[x] for x in f)
        elif m in (Detect, Segment, Pose):
            args.append([ch[x] for x in f])
            if m is Segment:
                args[2] = make_divisible(min(args[2], max_channels) * width, 8)
        elif m is RTDETRDecoder:  # special case, channels arg must be passed in index 1
            args.insert(1, [ch[x] for x in f])
        else:
            c2 = ch[f]

        m_ = nn.Sequential(*(m(*args) for _ in range(n))) if n > 1 else m(*args)  # module
        t = str(m)[8:-2].replace('__main__.', '')  # module type
        m.np = sum(x.numel() for x in m_.parameters())  # number params
        m_.i, m_.f, m_.type = i, f, t  # attach index, 'from' index, type
        if verbose:
            LOGGER.info(f'{i:>3}{str(f):>20}{n_:>3}{m.np:10.0f}  {t:<45}{str(args):<30}')  # print
        save.extend(x % i for x in ([f] if isinstance(f, int) else f) if x != -1)  # append to savelist
        layers.append(m_)
        if i == 0:
            ch = []
        ch.append(c2)
    return nn.Sequential(*layers), sorted(save)`

with these preceding changes, I finally modified the yaml file at
**ultralytics/cfg/models.v8/yolov8.yaml**

![yaml modification](https://github.com/ultralytics/ultralytics/assets/84653106/8bfa5735-744d-495f-afef-bd8450d649a3)

now I ran the training process
![train](https://github.com/ultralytics/ultralytics/assets/84653106/16aae30c-4d6d-4331-b91a-5110e437481b)
![result](https://github.com/ultralytics/ultralytics/assets/84653106/3980a9f9-fd3b-40aa-9d8d-7383ffbb2dff)

and I found out the layer structures printed did not change at all.

1. I wondered if the layers I added are not working properly so I changed the new part with some other existing convs and blocks(like C2F or other Conv layers) with wrong input and output sizes too. So I can see if it can cause error but the model training process still printed the same model structure as before.

2. As you can see from the picture of yaml file, I made each new layer arguments for 'from' to its previous(-1) and adjusted the last Detect layer according to it (it was originally 21 but I changed it to 24 so I can see if the values can really be passed by from these new layers. and it still prints 21 at the training process output)

3.  I checked some debugging lines at the forward methods of conv and block classes. others were regularly printing and forwarding the tensor values but my new conv and block does not seem to print anything. (I think it is not applied at all)

would there be anything missing here?

can you please let me understand how I can exactly change the structure of the model and affect the training process output so I can see whether it really changed or not?

Great Thank you for your interest and works

**Comment by glenn-jocher** [comment]
@jinyoonok2 hello! I'm glad to see your enthusiasm in customizing the YOLOv8 model architecture. 🛠 

It's important to ensure each new module is correctly registered within the codebase so the model definition in the YAML file is properly recognized. Here are the steps to debug your issue:

1. After creating new layers in the `conv.py` and `block.py` files, make sure that they are properly imported and accessible within the package.

2. In the `__init__.py` file, ensure that your new layers are imported without any namespace issues. These layers should then be visible when you list the contents of the package.

3. Within `tasks.py`, new layers need to be integrated in the `parse_model()` function so that they are properly constructed when the model is being built from the YAML definition.

If your layers are not being printed or invoked during training, double-check the following:

- YAML Configuration: Confirm that the new layers are correctly referenced in the `yolov8.yaml` configuration file. Take special care with the 'from' indices and layer names.

- Model Instantiation: Make sure that when the model is instantiated, the modified `parse_model()` with your custom layers is being called. 

- Activation Trace: Use Python's debugging tools or print statements to trace layer activation within the `forward()` method to ensure your new blocks are being executed.

If you've already performed these checks and the layers are still not executing during training, there may be an issue with how the layers are referenced or used within the model forward pass. Reviewing the model's forward definition and ensuring consistency between the layer definitions and their usage should help.

I hope this provides some guidance on how to proceed. Keep exploring and iterating, and good luck with your modifications! 🚀

**Comment by jinyoonok2** [comment]
> @jinyoonok2 hello! I'm glad to see your enthusiasm in customizing the YOLOv8 model architecture. 🛠
> 
> It's important to ensure each new module is correctly registered within the codebase so the model definition in the YAML file is properly recognized. Here are the steps to debug your issue:
> 
> 1. After creating new layers in the `conv.py` and `block.py` files, make sure that they are properly imported and accessible within the package.
> 2. In the `__init__.py` file, ensure that your new layers are imported without any namespace issues. These layers should then be visible when you list the contents of the package.
> 3. Within `tasks.py`, new layers need to be integrated in the `parse_model()` function so that they are properly constructed when the model is being built from the YAML definition.
> 
> If your layers are not being printed or invoked during training, double-check the following:
> 
> * YAML Configuration: Confirm that the new layers are correctly referenced in the `yolov8.yaml` configuration file. Take special care with the 'from' indices and layer names.
> * Model Instantiation: Make sure that when the model is instantiated, the modified `parse_model()` with your custom layers is being called.
> * Activation Trace: Use Python's debugging tools or print statements to trace layer activation within the `forward()` method to ensure your new blocks are being executed.
> 
> If you've already performed these checks and the layers are still not executing during training, there may be an issue with how the layers are referenced or used within the model forward pass. Reviewing the model's forward definition and ensuring consistency between the layer definitions and their usage should help.
> 
> I hope this provides some guidance on how to proceed. Keep exploring and iterating, and good luck with your modifications! 🚀

Hello @glenn-jocher , thank you for your reply! and I checked carefully about the settings again that were suggested.

While I am trying with various debugging lines, I found out that actually the model creation is not dont with the yolov8.yaml file I had modified

By that means, yolov8.yaml file in ultralytics/cfg/models/v8/yolov8.yaml did not interfere with any process of training while uploading the model and it was not even needed to run the training process

![without_yaml](https://github.com/ultralytics/ultralytics/assets/84653106/e694d635-db37-43b7-9a16-8ed695a05b48)

Above picture shows that I removed yolov8.yaml file from the v8 directory which I priorly modified before. so currently, system does not have model yaml file at all for the detection model. Now I tried to run the model.train() with the code at the right side.

![result1](https://github.com/ultralytics/ultralytics/assets/84653106/e220cd20-d23a-492e-8f10-c2f62c425001)
![result2](https://github.com/ultralytics/ultralytics/assets/84653106/72e5535f-1ff5-4c58-9ddc-c81ac68b1652)

and I found out the model just normally ran the training process with its original structure of layers even without the yaml file.

Therefore I concluded that the reason the changes I made did not affect the process is because the suggested yaml file does not interfere in the model construction process.

I checked some debugging lines for the parse_model() function as well,
the printed lines start with "This is cfg: ..." at the picture are the proof of those debugging lines, and I could see that it is called normally but I just found out that the yaml file is not the source of the model construction.

Can I ask if there is some other source that the YOLOv8 takes its model structure from except the yolov8.yaml file? if there is another one, is there also a way I can make it change its source to the yolov8.yaml file so I can apply my changes I made in the yolov8.yaml file?

Again, thank you for your time and attention

**Comment by glenn-jocher** [comment]
Hello @jinyoonok2!

It appears that your modification to the `yolov8.yaml` was not utilized during training, which could mean that another model configuration is being referenced.

To address this, it's crucial to ensure that:

1. You're modifying the correct `.yaml` file that the training script is configured to use. Sometimes, alternate config files are specified in the training code or command-line arguments.
2. When starting the training process, explicitly point to your modified `.yaml` file to ensure it is being loaded.

Please verify the path to the `.yaml` file in your training invocation. If you're using a command-line argument or a variable in your script to specify the model configuration, update it to point to your modified `yolov8.yaml`.

By explicitly directing the training routine to your custom `.yaml` file, you should be able to test your modified architecture. Good luck! 🚀

**Comment by jinyoonok2** [comment]
> Hello @jinyoonok2!
> 
> It appears that your modification to the `yolov8.yaml` was not utilized during training, which could mean that another model configuration is being referenced.
> 
> To address this, it's crucial to ensure that:
> 
> 1. You're modifying the correct `.yaml` file that the training script is configured to use. Sometimes, alternate config files are specified in the training code or command-line arguments.
> 2. When starting the training process, explicitly point to your modified `.yaml` file to ensure it is being loaded.
> 
> Please verify the path to the `.yaml` file in your training invocation. If you're using a command-line argument or a variable in your script to specify the model configuration, update it to point to your modified `yolov8.yaml`.
> 
> By explicitly directing the training routine to your custom `.yaml` file, you should be able to test your modified architecture. Good luck! 🚀

Wow thank you very much I can't believe it could be done this easy
it seems explicitly setting the yaml file for the model setting indeed solves the problem

`model = YOLO('yolov8s.yaml').load('yolov8s.pt')`

This gave me access to the modified model yaml file right away

I appreciate your time and replies!

**Comment by glenn-jocher** [comment]
Great to hear your issue is resolved, @jinyoonok2! Ensuring the right configuration file is used can often be the key. If you have any more questions or need further assistance as you continue exploring YOLOv8, don't hesitate to reach out. Happy coding! 🎉

**Comment by miguellpcs** [comment]
Hello all,
I replicated the steps described by @jinyoonok2 and it worked fine when adding the custom layers in the Head.

But when I tried to adding the custom layer to the backbone, by modifying the YAML file:
![image](https://github.com/ultralytics/ultralytics/assets/49253446/9b880f48-ecf1-4ee3-b1c4-2a58b1757659)


I got this error:

![image](https://github.com/ultralytics/ultralytics/assets/49253446/e6ae090d-0ebd-455f-989d-2600e762a074)

What else should I modify to be able to customize the backbone? 
Thanks ;)
